challenge_13/action_dispatch.rb

```diff
@@ -24,79 +24,61 @@ class Blog < ActiveRecord::Base; end
-    response.redirect('/show-data', 303)
-    response.finish
-  }
-  match '*path', via: :all, to: -> environment {
-    response = Rack::Response.new
-    response.write "Sorry, I don’t know what #{environment['PATH_INFO']} is"
-    response.content_type = 'text/plain'
-    response.status = 404
-    response.finish
-  }
-end
-
-class LoggerMiddleware
-  def initialize(app, string)
-    @app = app
-    @string = string
+class MyApp
+  def initialize
+    @router = ActionDispatch::Routing::RouteSet.new
+    draw_routes
   end
 
   def call(environment)
-    puts @string
-    @app.call(environment)
+    @router.call(environment)
   end
-end
 
-class CharsetMiddleware
-  def initialize(app, charset: 'UTF-8')
-    @app = app
-    @charset = charset
-  end
+  private
 
-  def call(environment)
-    response = Rack::Response[*@app.call(environment)]
-    response.content_type = "#{response.content_type}; charset=#{@charset}"
-    response.finish
+  def draw_routes
+    @router.draw do
+      get '/', to: -> environment {
+        response = Rack::Response.new
+        response.write '<p><strong>Submit a new Blog Post!</p></strong>'
+        response.write "<form method='post' enctype='application/x-www-form-urlencoded' action='/create-post'>"
+        response.write "<p><label>Blog Title: <input name='title'></label></p>"
+        response.write "<p><label>Content: <textarea name='content'></textarea></label></p>"
+        response.write '<p><button>Submit post</button></p>'
+        response.write '</form>'
+        response.content_type = 'text/html'
+        response.finish
+      }
+      get '/show-data', to: -> environment {
+        response = Rack::Response.new
+        response.content_type = 'text/html'
+        response.finish do
+          response.write '<ul>'
+    
+          Blog.all.each do |blog|
+            response.write '<li>'
+            response.write "<strong>Title: #{CGI.escape_html(blog.title)}</strong>, Content: #{CGI.escape_html(blog.content)}"
+            response.write '</li>'
+          end
+          response.write '</ul>'
+        end
+      }
+      post 'create-post', to: -> environment {
+        response = Rack::Response.new
+        request = Rack::Request.new(environment)
+        puts 'Got a new POST request!'
+    
+        Blog.create!(request.params)
+        response.redirect('/show-data', 303)
+        response.finish
+      }
+      match '*path', via: :all, to: -> environment {
+        response = Rack::Response.new
+        response.write "Sorry, I don’t know what #{environment['PATH_INFO']} is"
+        response.content_type = 'text/plain'
+        response.status = 404
+        response.finish
+      }
+    end
   end
 end
-
-app = lambda { |environment| router.call(environment) }
-app = CharsetMiddleware.new(app)
-app = LoggerMiddleware.new(app, 'Rack app got a request!')
-
-# Run the Puma app server with our web application
-Rack::Handler::Puma.run(app, Port: 1234, Verbose: true)
```

Comment:
As you noted, we’ve lost the ability to control the port number and verbosity ourselves now that we’re no longer starting Puma, but that’s kind of a good thing because instead it’s up to the person _running the application_:

```
% puma --help
puma <options> <rackup file>
    -b, --bind URI                   URI to bind to (tcp://, unix://, ssl://)
        --bind-to-activated-sockets [only]
                                     Bind to all activated sockets
    -C, --config PATH                Load PATH as a config file
        --control-url URL            The bind url to use for the control server. Use 'auto' to use temp unix server
        --control-token TOKEN        The token to use as authentication for the control server
        --debug                      Log lowlevel debugging information
        --dir DIR                    Change to DIR before starting
    -e, --environment ENVIRONMENT    The environment to run the Rack app on (default development)
    -f, --fork-worker=[REQUESTS]     Fork new workers from existing worker. Cluster mode only
                                     Auto-refork after REQUESTS (default 1000)
    -I, --include PATH               Specify $LOAD_PATH directories
    -p, --port PORT                  Define the TCP port to bind to
                                     Use -b for more advanced options
        --pidfile PATH               Use PATH as a pidfile
        --preload                    Preload the app. Cluster mode only
        --prune-bundler              Prune out the bundler env if possible
        --extra-runtime-dependencies GEM1,GEM2
                                     Defines any extra needed gems when using --prune-bundler
    -q, --quiet                      Do not log requests internally (default true)
    -v, --log-requests               Log requests as they occur
    -R, --restart-cmd CMD            The puma command to run during a hot restart
                                     Default: inferred
    -S, --state PATH                 Where to store the state details
    -t, --threads INT                min:max threads to use (default 0:16)
        --early-hints                Enable early hints support
    -V, --version                    Print the version information
    -w, --workers COUNT              Activate cluster mode: How many worker processes to create
        --tag NAME                   Additional text to display in process listing
        --redirect-stdout FILE       Redirect STDOUT to a specific file
        --redirect-stderr FILE       Redirect STDERR to a specific file
        --[no-]redirect-append       Append to redirected files
    -h, --help                       Show help
```

So a user who doesn’t care can type simply `puma` and get the defaults, but another can say `puma --port 1234 --log-requests` if they want the previous behaviour.

Reply:
Oh yay!!! 🙌